primitives/src/crh/mod.rs

```diff
@@ -38,5 +38,11 @@ pub trait BatchFieldBasedHash {
     type Data: Field;
     type Parameters: FieldBasedHashParameters<Fr = Self::Data>;
 
-    fn batch_evaluate_2_1(input_array: &mut[Self::Data]);
-}
\ No newline at end of file
+    fn batch_evaluate(input_array: &mut[Self::Data], output_array: &mut[Self::Data]);
+
+    fn merkle_tree(input_vec: &mut[Self::Data], output_vec: &mut[Self::Data], input_size: usize);
```

Comment:
I understand the rationale behind this and we should think more about it. So conceptually, according to me, it's not the best choice to let the caller pass the array that will contain the solution, because then he needs to worry about its size and we need to make additional internal checks so to be sure that the size is right. 
However, in the context of the highly optimized batched Merkle Tree implementation, returning a new array implies a Copy and, therefore, performance degradation.
We should exstabilish if it makes sense to have a BatchFieldBasedHash standalone, i.e. if there exist other real-world common applications of it apart from MerkleTree, and in this case it makes sense to return a new array from the function and separate the traits BatchFieldBasedHash and BatchedFieldBasedMerkleTree, or not, and in this case keep everything in a single BatchedFieldBasedMerkleTree.

Reply:
Moved to new trait.